Use ScreenCoordsXY for windows/Map

tupaschoal opened the pull request:
- Refactor `place_park_entrance_get_map_position()` to receive `ScreenCoordsXY` and return `CoordsXYZD`
- Refactor `sub_68A15E()` to receive `ScreenCoordsXY` and return `CoordsXY`. Some of its usage are narrowing the result on purpose, as same variable is used for contexts where `int16_t` is still a hard requirement.

duncanspumpkin commented on src/openrct2-ui/windows/Map.cpp:
```diff
@@ -1189,89 +1189,92 @@ static void window_map_set_land_rights_tool_update(int32_t x, int32_t y)
-    if (*mapZ == 0)
+    CoordsXYZD parkEntranceMapPosition{};
+    const CoordsXY mapCoords = sub_68A15E(screenCoords);
+    parkEntranceMapPosition = { mapCoords.x, mapCoords.y, parkEntranceMapPosition.z, parkEntranceMapPosition.direction };
```
Keep the initialised values as explicit 0 and invalid direction

duncanspumpkin commented on src/openrct2-ui/windows/Map.cpp:
```diff
@@ -1189,89 +1189,92 @@ static void window_map_set_land_rights_tool_update(int32_t x, int32_t y)
+    if (parkEntranceMapPosition.x == LOCATION_NULL)
+        return parkEntranceMapPosition;
+
+    auto surfaceElement = map_get_surface_element_at(parkEntranceMapPosition.x >> 5, parkEntranceMapPosition.y >> 5);
```
I think the is an override that takes a coordsxy

tupaschoal replied:
There is, thanks!

duncanspumpkin commented on src/openrct2-ui/windows/Map.cpp:
```diff
@@ -1189,89 +1189,92 @@ static void window_map_set_land_rights_tool_update(int32_t x, int32_t y)
+
+    auto surfaceElement = map_get_surface_element_at(parkEntranceMapPosition.x >> 5, parkEntranceMapPosition.y >> 5);
+    parkEntranceMapPosition.z = surfaceElement->GetWaterHeight();
+    if (parkEntranceMapPosition.z == 0)
```
All of this z code should be setting the value in terms of big coordinates not micro. Will need the appropriate *8's

duncanspumpkin commented on src/openrct2-ui/windows/Map.cpp:
```diff
@@ -1189,89 +1189,92 @@ static void window_map_set_land_rights_tool_update(int32_t x, int32_t y)
-    gMapSelectArrowDirection = direction;
+    gMapSelectArrowPosition.x = parkEntrancePosition.x;
+    gMapSelectArrowPosition.y = parkEntrancePosition.y;
+    gMapSelectArrowPosition.z = parkEntrancePosition.z * 16;
```
Conversion from micro. Once anniversary changes are made this needs modified

duncanspumpkin commented on src/openrct2-ui/windows/Map.cpp:
```diff
@@ -1189,89 +1189,92 @@ static void window_map_set_land_rights_tool_update(int32_t x, int32_t y)
     park_entrance_remove_ghost();
-    park_entrance_place_ghost(mapX, mapY, mapZ, direction);
+    park_entrance_place_ghost(
+        parkEntrancePosition.x, parkEntrancePosition.y, parkEntrancePosition.z, parkEntrancePosition.direction);
```
Place ghost expects a micro z

duncanspumpkin commented on src/openrct2-ui/windows/Map.cpp:
```diff
@@ -1302,16 +1305,15 @@ static void window_map_set_peep_spawn_tool_update(int32_t x, int32_t y)
     {
-        money32 price = place_park_entrance(mapX, mapY, mapZ, direction);
+        money32 price = place_park_entrance(
+            parkEntrancePosition.x, parkEntrancePosition.y, parkEntrancePosition.z, parkEntrancePosition.direction);
```
Micro z

tupaschoal commented:
I've fixed the issues you mentioned @duncanspumpkin and also rebased to solve conflicts. Please double check the micro/big z changes to make sure I got it right

duncanspumpkin commented on src/openrct2-ui/windows/Map.cpp:
```diff
@@ -1188,89 +1188,92 @@ static void window_map_set_land_rights_tool_update(int32_t x, int32_t y)
         if ((surfaceElement->GetSlope() & TILE_ELEMENT_SLOPE_ALL_CORNERS_UP) != 0)
         {
-            (*mapZ)++;
+            parkEntranceMapPosition.z++;
```
+= 16

tupaschoal replied:
```suggestion
            parkEntranceMapPosition.z += 16;
```

duncanspumpkin commented on src/openrct2-ui/windows/Map.cpp:
```diff
@@ -1188,89 +1188,92 @@ static void window_map_set_land_rights_tool_update(int32_t x, int32_t y)
             if (surfaceElement->GetSlope() & TILE_ELEMENT_SLOPE_DOUBLE_HEIGHT)
             {
-                (*mapZ)++;
+                parkEntranceMapPosition.z++;
```
16

duncanspumpkin commented on src/openrct2-ui/windows/Map.cpp:
```diff
@@ -1188,89 +1188,92 @@ static void window_map_set_land_rights_tool_update(int32_t x, int32_t y)
-    place_park_entrance_get_map_position(x, y, &mapX, &mapY, &mapZ, &direction);
-    if (mapX == (int16_t)-1)
+    CoordsXYZD parkEntrancePosition = place_park_entrance_get_map_position(screenCoords);
+    if (parkEntrancePosition.x == -1)
```
Location null

tupaschoal replied:
```suggestion
    if (parkEntrancePosition.x == LOCATION_NULL)
```

tupaschoal commented on src/openrct2-ui/windows/Map.cpp:
```diff
@@ -1188,89 +1188,92 @@ static void window_map_set_land_rights_tool_update(int32_t x, int32_t y)
             if (surfaceElement->GetSlope() & TILE_ELEMENT_SLOPE_DOUBLE_HEIGHT)
             {
-                (*mapZ)++;
+                parkEntranceMapPosition.z++;
```
```suggestion
                parkEntranceMapPosition.z += 16;
```

Gymnasiast replied:
Just to check: normally you would use a factor of 8 when converting from small to big coordinates.

Was parkEntranceMapPosition.z using half-tile units before?

tupaschoal commented:
Solved the pending issues, should be alright now

Gymnasiast commented:
Some of the diff looks weird, but that it caused by the park entrance structures using coordinates that are neither big nor small, but an oddball mixture. That will need to be refactored, but not in this PR.
